augur/metadata_file.py

```diff
@@ -0,0 +1,89 @@
+import functools
+from os import access, R_OK
+import pandas
+
+
+class MetadataFile:
+    """
+    Represents a CSV or TSV file containing metadata
+
+    The file must contain exactly one of a column named `strain` or `name`,
+    which is used to match metadata with samples.
+    """
+
+    def __init__(self, fname, query):
+        self.fname = fname
+        self.query = query
+
+        self.key_type = self.find_key_type()
+
+        self.check_file()
+
+    def check_file(self):
+        if not access(self.fname, R_OK):
+            raise ValueError(
+                f"Metadata file {self.fname} does not exist or is not readable"
+            )
+
+    def read(self):
+        meta_dict = {}
+        for row in self.metadata:
+            key_val = row[self.key_type]
+
+            if key_val in meta_dict:
+                raise ValueError(
+                    f"Duplicate {self.key_type} in metadata files: `{key_val}`"
+                )
+
+            meta_dict[key_val] = row
+
+        return meta_dict, self.columns
+
+    @property
+    def metadata(self):
+        """
+        Return list of dicts representing the metadata in the file.
+
+        If a query was supplied, apply it.
+        """
+
+        metadata = self.parse_file()
+
+        if self.query:
+            try:
+                metadata = metadata.query(self.query)
+            except ValueError:
+                raise ValueError(
+                    f"Error applying pandas query to metadata: `{self.query}`"
+                )
+
+        return [row[1].to_dict() for row in metadata.iterrows()]
+
+    @property
+    @functools.lru_cache()
+    def columns(self):
+        return list(self.parse_file().columns)
+
+    def find_key_type(self):
+        if "strain" not in self.columns and "name" not in self.columns:
+            raise ValueError(
+                f"Metadata file {self.fname} does not contain `name` or `strain`"
+            )
+
+        if "strain" in self.columns and "name" in self.columns:
+            raise ValueError(
+                f"Metadata file {self.fname} contains both `name` and `strain`"
+            )
```

Comment:
I agree this is a weird edge case that should be handled better than the code in `master`. The legacy behavior is to prefer the `strain` column over `name` if both are defined. I'm not sure the presence of both should be treated as an error, but it might warrant a warning.

@jameshadfield @emmahodcroft @tsibley Do y'all have thoughts about this behavior?

Reply:
I think a warning is warranted, but to avoid breaking existing builds, I wouldn't make this an error condition. It doesn't seem severe enough to warrant it.

Reply:
Changed to emit a warning and prefer `strain`.